contrib/jms/src/main/java/com/google/cloud/dataflow/contrib/jms/JmsIO.java

```diff
@@ -0,0 +1,532 @@
+             */
+            public Bound<T> connectionFactory(ConnectionFactory connectionFactory) {
+                return new Bound<T>(name, connectionFactory, queue, topic, coder, maxNumMessages);
+            }
+
+            /**
+             * Returns a transform that's like this one but that reads
+             * from the specified JMS queue.
+             *
+             * @param queue the JMS queue
+             * @return the Bound
+             */
+            public Bound<T> queue(String queue) {
+                return new Bound<T>(name, connectionFactory, queue, topic, coder, maxNumMessages);
+            }
+
+            /**
+             * Returns a transform that's like this one but that reads
+             * from the specified JMS topic.
+             *
+             * @param topic the JMQ topic
+             * @return the Bound
+             */
+            public Bound<T> topic(String topic) {
+                return new Bound<T>(name, connectionFactory, queue, topic, coder, maxNumMessages);
+            }
+
+            /**
+             * Returns a transform that's like this one but that uses the
+             * given {@link Coder} to decode each record into a value of
+             * type {@code X}.
+             *
+             * @param coder the coder to use
+             * @return the Bound
+             */
+            public <X> Bound<X> withCoder(Coder<X> coder) {
+                return new Bound<>(name, connectionFactory, queue, topic, coder, maxNumMessages);
+            }
+
+            /**
+             * Return a transform that's like this one but will only read up
+             * to the specified maximum number of messages from the JMS broker.
+             * The transform produces a <i>bounded</i> {@link PCollection}.
+             *
+             * @param maxNumMessages the max number messages limit to consume
+             * @return the Bound
+             */
+            public Bound<T> maxNumMessages(int maxNumMessages) {
+                return new Bound<T>(name, connectionFactory, queue, topic, coder, maxNumMessages);
+            }
+
+            @Override
+            public PCollection<T> apply(PInput input) {
+                if (connectionFactory == null) {
+                    throw new IllegalArgumentException("need to set connection factory " +
+                            "for a JmsIO.Read transform");
+                }
+                if (queue == null && topic == null) {
+                    throw new java.lang.IllegalStateException("need to set destination " +
+                            "(queue or topic) for a JmsIO.Read transform");
+                }
+
+                boolean boundedOutput = getMaxNumMessages() > 0;
+
+                if (boundedOutput) {
+                    return input.getPipeline().begin()
+                            .apply(Create.of((Void) null)).setCoder(VoidCoder.of())
+                            .apply(ParDo.of(new JmsReader())).setCoder(coder);
```

Comment:
Hmm -- I expected to see you using both the `BoundedSource` and `UnboundedSource` APIs, rather than this approach. Specifically, this approach looks like it won't scale to more than 1 worker.

It looks like you based this implementation on `PubSubIO`? If so, that is very unfortunate -- the version of pub/sub that is in the SDK is not meant for production use, but rather only for testing with the DirectPipelineRunner.  https://issues.apache.org/jira/browse/BEAM-53